client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -907,8 +908,43 @@ public final void shutdown() {
     @Override
     public final void shutdown(final Handler<AsyncResult<Void>> completionHandler) {
 
-        if (shuttingDown.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
+        closeConnection(completionHandler, true);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect() {
+
+        disconnect(disconnectResult -> {
+            if (disconnectResult.succeeded()) {
+                LOG.info("successfully disconnected from the server");
+            } else {
+                LOG.error("could not disconnect from the server", disconnectResult.cause());
+            }
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+        closeConnection(completionHandler, false);
+    }
+
+    //-----------------------------------< private methods >---
+
+    private void closeConnection(final Handler<AsyncResult<Void>> completionHandler, final Boolean isShuttingDown) {
+
+        Objects.requireNonNull(completionHandler, "Result handler cannot be null");
+        AtomicBoolean flag = (isShuttingDown) ? shuttingDown : disconnecting;
```

Comment:
what if *isShuttingDown* is `null`
flag should be declared final

Reply:
but the flag is already declared final or?